# BioinformaticsDemo
Independent study from a uDemy bioinformatics course, this script uses Biopython functions to analyze SARSCOV2 genomes

With Biopython dependences some 'general use' functions are defined here: Protein_Estimator() and CODON_READER() - which can create potential amino acids sequences from any given nucleotide sequence.

In this script Protein_Estimator() creates potential protein candidates from the spike gene segment of the SARSCOV2 genomes. Each set of variant spike Amino Acid sequences is then compared to the alpha spike sequences using PROTEIN_COMPARER(). Outputs were validated by BLASTing manually:

DELTA_LIST[10]['SEQA'] corresponds to NCBI YP_009724390.1\n
DELTA_LIST[10]['SEQB'] corresponds to NCBI QXQ22678.1\n
MU_LIST[8]['SEQB'] corresponds to NCBI QZW69948.1\n
OM_LIST[8]['SEQB'] corresponds to NCBI UFO69279.1.\n

Each of these 4 AA sequences are then compared to the AA sequence generated by Pfizer mRNA using VAX_DRIFT_REPORT(). This function prints identity scores and the results demonstrate viral mutation to evade induced immunity.
