# BioinformaticsDemo
Independent study from a uDemy bioinformatics course, these scripts uses Biopython functions to analyze SARS-COV2 genomes

With Biopython dependences some 'general use' functions are defined in COVID19informaticsDEMO.py: Protein_Estimator() and CODON_READER() - which can create potential amino acids sequences from any given nucleotide sequence.

COVID19informaticsDEMO.py inputs 4 genomes (the original 'alpha' and 3 variants). Protein_Estimator() creates potential protein candidates from the spike gene segment of each SARS-COV2 genomes. Each set of variant spike Amino Acid sequences is then compared to the 'alpha' spike sequence using PROTEIN_COMPARER(). Outputs were validated by BLASTing manually:

DELTA_LIST[10]['SEQA'] corresponds to NCBI YP_009724390.1

DELTA_LIST[10]['SEQB'] corresponds to NCBI QXQ22678.1

MU_LIST[8]['SEQB'] corresponds to NCBI QZW69948.1

OM_LIST[8]['SEQB'] corresponds to NCBI UFO69279.1

Each of these 4 AA sequences are then compared to the AA sequence generated by Pfizer mRNA using VAX_DRIFT_REPORT(). This function prints identity scores and the results demonstrate viral mutation to evade induced immunity.

Additional script SARSCOV2InformaticsCheck.py inputs 8 different genome samples and outputs FASTA files for specific genes (Spike, Membrane, Envelope) for BLAST analysis of nucleotide-level changes.
